# Favourite Free Apps in App Store and Google Play Store

In this analysis, we are going to understand what kind of apps in App Store and Google Play Store are loved by user. The result of this analysis will be used to make recomendation to a company that builds Android and iOS mobile apps. 

In this case, the company ony build apps that are free to download and install and the main source of revenue consists of in-app ads. This means that the revenue of the company for any given app is mostly influenced by the number of users who use their app. The more users that see and attracted to the ads, the more revenue the company will get. This company is also targeted an English speaking audience.

## Data Sets

To achieve the goal of this analysis, we should collect data from App Store and Google Play. According to [Statista](https://www.statista.com/statistics/276623/number-of-apps-available-in-leading-app-stores/), there were approximately 2 million iOS apps available on the App Store and 2.1 million Android apps on Google Play. Collecting all of the apps data requires a significant amount of time and money. So, we will try to get the sample of the data. Luckily, we can find two datasets in the internet that seem suitable for our goals.
1. [App Store](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home). This dataset are collected by Ramanathan in July 2017. This dataset contain about 7200 apps from App Store.
2. [Google Play](https://www.kaggle.com/lava18/google-play-store-apps/home). This dataset are collected by Lavanya Gupta. This dataset contain about 10,000 apps from Google Play.

In [8]:
#Import Dataset to python
from csv import reader

#App Store dataset
opened_data=open('AppleStore.csv', encoding="utf8")
read_data=reader(opened_data)
appstore=list(read_data)
ios_header=appstore[0]
ios=appstore[1:]

#Google Play Store dataset
opened_data=open('googleplaystore.csv', encoding="utf8")
read_data=reader(opened_data)
googleplay=list(read_data)
android_header=googleplay[0]
android=googleplay[1:]

## Explore Data Set

To explore the data easier, we make a function called explore_data(). This function can help us to read some row from the data set in more readable way. It also have an optional parameter to show the number of rows and column from the data set.

In [9]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

Let's take a look at the App Store data set.

In [10]:
#Show App Store data set header
print(ios_header, "\n")

#Show first three row and the number of rows and columns from the data set
explore_data(ios, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


From information above, we know that App Store dataset has 7197 rows and 16 column. The columns that could help us with our analysis are track_name, price, rating_count_tot, rating_count_ver, user_rating, user_rating_ver, and prime_genre. The description of the column name can be seen [here](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home).

Now let's take a look on Google Play Store data set

In [32]:
#Show Google Play Store data set header
print(android_header, "\n")

#Show first three row and the number of rows and columns from the data set
explore_data(android, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


From information above, we know that Google Play Store dataset has 10841 rows and 13 column. The columns that could help us with our analysis are App, Category, Rating, Install, Type, Price, and Genres. The description of the column name can be seen [here](https://www.kaggle.com/lava18/google-play-store-apps).

## Deleting wrong data

The Google Play data set has a dedicated discussion section, and we can see that one of the discussions describes an error for row 10472.

In [11]:
print(android_header,"\n")
print(android[10472])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


We can see that the rating of 'Life Made WI-Fi Touchscreen Photo Frame' app is 19. It definetly incorrect because the maximum rating in Google Play Store is 5. So, we will delete this row.

In [12]:
print(len(android))
del android[10472] #don't run more than once
print(len(android))

10841
10840


the data in row 10472 has been deleted

## Removing Duplicate Entries

In the discussion section, we also found one discussion about duplicate entries for the same app. For example, Instagram has four entries:

In [13]:
print(android_header)
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


Now, we are gonna count the number of duplicate app in Google Play data set.

In [14]:
duplicate_apps=[]
unique_apps=[]

for app in android:
    name=app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

print('Number of duplicate apps:', len(duplicate_apps),"\n")
print('Example of duplicate apps:', duplicate_apps[:15])

Number of duplicate apps: 1181 

Example of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


In the analysis, we don't want to count certain apps more than once. So we need to remove the duplicate entries and keep only one entry per app. The easiest way is remove the duplicates randomly, but we could probably find a better way.

From the example for Instagram app, we can see that the main difference happens on the fourth position of each row which corresponds to the number of reviews. The different numbers show the data was collected at different times. So, we will use the heighest number of reviews because it present the recent data should be. Then we will remove the other entries for any given app.

In [15]:
review_max={}

for app in android:
    name=app[0]
    n_review=float(app[3])
    if name in review_max and n_review > review_max[name]:
        review_max[name]=n_review
    elif name not in review_max:
        review_max[name]=n_review

In the previous code, we know that the number of duplicate apps are 1181 and the length of our data set are 10840. So, after remove the duplicates, the length of our data set should be the difference between  the initial length of our data set minus the number of duplicate apps.

In [16]:
print('Expected length:', len(android)-len(duplicate_apps))
print('Actual length:', len(review_max))

Expected length: 9659
Actual length: 9659


Now, we will remove the duplicate rows:

In [17]:
android_clean=[]
already_added=[]

for app in android:
    name=app[0]
    n_review=float(app[3])
    if n_review==review_max[name] and name not in already_added:
        android_clean.append(app)
        already_added.append(name)
    

In [18]:
len(android_clean)

9659

## Removing Non-English Apps

English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation mark (., !, ?, ;), and another symbols (+, /). Behind the scenes, each character has a corresponding number associated with it. For example, the corresponding number for character "a" is 97, for character "A" is 65, and character '爱' is 29,233. We can get the corresponding number of each character using the ord() build-in function.

The numbers corresponding to the characters we commonly use in an English text are all in the range 0 to 127, according to the ASCII (American Standard Code for Information Interchange) system. So, we will make a function that detects whether an app name is in English or not using this criteria.

In [19]:
def is_english(string):
    for character in string:
        if ord(character) > 127:
            return False
    return True

Checking the function

In [20]:
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


We can see that the function couldn't correctly identify certain English app name like 'Docs To Go™ Free Office Suite' and 'Instachat 😜'. This is because emojis and characters like '™' fall outside the ASCII range and have corresponding numbers over 127.

To minimize the data loss, we will only remove an app if its name has more three characters with corresponding numbers falling outside the ASCII range.

In [21]:
def is_english(string):
    non_ASCII=0
    for character in string:
        if ord(character) > 127:
            non_ASCII+=1
    if non_ASCII > 3:        
        return False
    else:    
        return True

Checking the function

In [22]:
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
True
True


We can see that the function can identify whether it is english name or not more correctly now. Let's implement this function to our dataset.

In [23]:
android_english=[]
ios_english=[]

for app in android_clean:
    name=app[0]
    if is_english(name):
        android_english.append(app)

for app in ios:
    name=app[1]
    if is_english(name):
        ios_english.append(app)

In [24]:
print(android_header)
print(explore_data(android_english,0,3,True))
print('\n')
print(ios_header)
print(explore_data(ios_english,0,3,True))

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13
None


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
['284882215

From the result above, we can see that the data left for Google Play Store app is 9,614 and for App Store is 6,183.

## Isolating the Free Apps

Because the company only build apps that are free to download and install. So, we will isolating the free apps because we just need that data.

In [25]:
android_free=[]
ios_free=[]

for app in android_english:
    if app[7]=='0':
        android_free.append(app)

for app in ios_english:
    if app[4]=='0.0':
        ios_free.append(app)

In [26]:
print(len(android_free))
print(len(ios_free))

8864
3222


From the result above, we can see that the data left for Google Play Store app is 9,614 and for App Store is 6,183.

## Most Common Apps by Genre

To identify the most common apps by genre, the columns that might be useful for us are the prime_genre column of App Store and the Geres and Category columns of Google Play Store. Now, we will build two function that we can use to analyze the frequency table. One function to generate frequency tables that show percentage and another function to display the percentages in a descanding order.

In [34]:
def freq_table(dataset, index):
    total_apps = {}
    for app in dataset:
        criteria = app[index]
        if criteria in total_apps:
            total_apps[criteria]+=1
        else:
            total_apps[criteria]=1
    percent_apps = {}
    for key in total_apps:
        percent_apps[key]=total_apps[key]/len(dataset)*100
    return percent_apps

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        as_tuple = (table[key], key)
        table_display.append(as_tuple)
    table_sorted = sorted(table_display, reverse = True)
    for a_tuple in table_sorted:
        print(a_tuple[1], ":", a_tuple[0])
    

In [35]:
display_table(ios_free, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


In [36]:
display_table(android_free, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [37]:
display_table(android_free, -4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

## Most Popular Apps by Genre on the App Store

In [45]:
genres = freq_table(ios_free, -5)

for genre in genres:
    total = 0
    len_genre = 0
    for app in ios_free:
        genre_app = app[-5]
        if genre_app == genre:
            total += float(app[5])
            len_genre += 1
    avg_rating = total/len_genre
    print(genre, ":", avg_rating)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


## Most Popular Apps by Genre on Google Play

In [58]:
categories = freq_table(android_free, 1)

for category in categories:
    total = 0
    len_category = 0
    for app in android_free:
        category_app = app[1]
        if category_app == category:
            num_install = app[5]
            num_install = num_install.replace("+","")
            num_install = num_install.replace(",","")
            total += float(num_install)
            len_category += 1
    avg_install = total/len_category
    print(category, ":", avg_install)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_